src/async-wrap.cc

```diff
@@ -194,52 +196,46 @@ Local<Value> AsyncWrap::MakeCallback(const Local<Function> cb,
     }
   }
 
-  TryCatch try_catch(env()->isolate());
-  try_catch.SetVerbose(true);
-
   if (has_domain) {
     Local<Value> enter_v = domain->Get(env()->enter_string());
     if (enter_v->IsFunction()) {
-      enter_v.As<Function>()->Call(domain, 0, nullptr);
-      if (try_catch.HasCaught())
-        return Undefined(env()->isolate());
+      if (enter_v.As<Function>()->Call(domain, 0, nullptr).IsEmpty()) {
+        FatalError("node::AsyncWrap::MakeCallback",
+                   "domain enter callback threw, please report this");
+      }
     }
   }
 
   if (ran_init_callback() && !pre_fn.IsEmpty()) {
-    try_catch.SetVerbose(false);
-    pre_fn->Call(context, 0, nullptr);
-    if (try_catch.HasCaught())
+    if (pre_fn->Call(context, 0, nullptr).IsEmpty())
       FatalError("node::AsyncWrap::MakeCallback", "pre hook threw");
-    try_catch.SetVerbose(true);
   }
 
   Local<Value> ret = cb->Call(context, argc, argv);
 
-  if (try_catch.HasCaught()) {
-    return Undefined(env()->isolate());
-  }
-
   if (ran_init_callback() && !post_fn.IsEmpty()) {
-    try_catch.SetVerbose(false);
-    post_fn->Call(context, 0, nullptr);
-    if (try_catch.HasCaught())
+    // TODO(trevnorris): Error reporting in async wrap post callback.
+    if (post_fn->Call(context, 0, nullptr).IsEmpty())
       FatalError("node::AsyncWrap::MakeCallback", "post hook threw");
-    try_catch.SetVerbose(true);
+  }
+
+  if (ret.IsEmpty()) {
```

Comment:
Verified this by placing the early return after the domain exit call and tested with the following:
```js
'use strict';

const domain = require('domain');
const d = domain.create();
const async_wrap = process.binding('async_wrap');

async_wrap.setupHooks(function() {
}, function() {
  process._rawDebug('before');
}, function() {
  process._rawDebug('after');
});
async_wrap.enable();

d.on('error', function() {
  process._rawDebug('d errored');
});

d.run(function() {
  require('crypto').randomBytes(1024, function() {
    throw new Error('crap');
  });
});
```